protocols/relay/src/behaviour/handler.rs

```diff
@@ -716,6 +741,21 @@ impl ConnectionHandler for Handler {
             }
         }
 
+        // Send stop commands
+        if let Poll::Ready(Some(result)) = self.outbound_stop_futs.poll_next_unpin(cx) {
+            return match result {
+                Ok((event, receiver)) => {
+                    if let Some(rx) = receiver {
+                        self.alive_lend_out_substreams.push(rx);
+                    }
+                    Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(event))
+                }
+                Err(e) => Poll::Ready(ConnectionHandlerEvent::Close(StreamUpgradeError::Apply(
+                    Either::Right(e),
+                ))),
+            };
+        }
```

Comment:
I think we can slightly simplify this:

Instead of returning the oneshot from the future, we can construct the oneshot early within `send_stop_message_and_process_result` and directly push it into `self.alive_lend_out_substreams`.

If the outbound-stop-future fails, the other end of the oneshot will be dropped and we will automatically clean this up. Whilst we are negotiating the outbound stream, it is still alive so I don't think there is any harm in constructing it early.

Reply:
> I think we can slightly simplify this:
> 
> Instead of returning the oneshot from the future, we can construct the oneshot early within `send_stop_message_and_process_result` and directly push it into `self.alive_lend_out_substreams`.
> 
> If the outbound-stop-future fails, the other end of the oneshot will be dropped and we will automatically clean this up. Whilst we are negotiating the outbound stream, it is still alive so I don't think there is any harm in constructing it early.

When I try to change `self.alive_lend_out_substreams` i get the following:
```
error: future cannot be sent between threads safely
   --> protocols/relay/src/behaviour/handler.rs:440:9
    |
440 | /         async move {
441 | |             substream.send(msg).await?;
442 | |
443 | |             let proto::StopMessage {
...   |
515 | |             })
516 | |         }
    | |_________^ future created by async block is not `Send`
    |
    = help: the trait `Sync` is not implemented for `(dyn futures::Future<Output = Result<behaviour::handler::Event, outbound_stop::FatalUpgradeError>> + std::marker::Send + 'static)`
note: captured value is not `Send` because `&` references cannot be sent unless their referent is `Sync`
   --> protocols/relay/src/behaviour/handler.rs:506:13
    |
506 |             self.alive_lend_out_substreams.push(rx);
    |             ^^^^ has type `&behaviour::handler::Handler` which is not `Send`, because `behaviour::handler::Handler` is not `Sync`
note: required by a bound in `futures::FutureExt::boxed`
   --> /Users/denisgarus/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.28/src/future/future/mod.rs:520:23
    |
520 |         Self: Sized + Send + 'a,
    |                       ^^^^ required by this bound in `FutureExt::boxed`
 ```
That is why I return the `oneshot` from the future.
Could you please tell me the right way to fix this?

Reply:
You can't modify it inside the future, you'd have to construct the oneshot within the function but before the `async` block. Then you can push `rx` into the `self.alive_lend_out_substreams` and move `tx` into the future :)